src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -968,13 +976,65 @@ void RaftReplDev::cp_flush(CP*) {
         // Not dirtied since last flush ignore
         return;
     }
+
+    std::unique_lock lg{m_sb_mtx};
     m_rd_sb->compact_lsn = clsn;
-    m_rd_sb->commit_lsn = lsn;
+    m_rd_sb->durable_commit_lsn = lsn;
     m_rd_sb->checkpoint_lsn = lsn;
     m_rd_sb->last_applied_dsn = m_next_dsn.load();
     m_rd_sb.write();
     m_last_flushed_commit_lsn = lsn;
 }
 
 void RaftReplDev::cp_cleanup(CP*) {}
+
+void RaftReplDev::on_log_found(logstore_seq_num_t lsn, log_buffer buf, void* ctx) {
+    // apply the log entry if the lsn is between checkpoint lsn and durable commit lsn
+    if (m_rd_sb->checkpoint_lsn >= lsn || m_rd_sb->durable_commit_lsn <= lsn) { return; }
+
+    // 1. Get the log entry and prepare rreq
+    nuraft::log_entry const* lentry = r_cast< nuraft::log_entry const* >(buf.bytes());
+    repl_journal_entry* jentry = r_cast< repl_journal_entry* >(lentry->get_buf().data_begin());
+    RELEASE_ASSERT_EQ(jentry->major_version, repl_journal_entry::JOURNAL_ENTRY_MAJOR,
+                      "Mismatched version of journal entry received from RAFT peer");
+
+    RD_LOGT("Raft Channel: Applying Raft log_entry upon recovery: server_id={}, term={}, journal_entry=[{}] ",
+            jentry->server_id, lentry->get_term(), jentry->to_string());
+
+    auto entry_to_hdr = [](repl_journal_entry* jentry) {
+        return sisl::blob{uintptr_cast(jentry) + sizeof(repl_journal_entry), jentry->user_header_size};
+    };
+
+    auto entry_to_key = [](repl_journal_entry* jentry) {
+        return sisl::blob{uintptr_cast(jentry) + sizeof(repl_journal_entry) + jentry->user_header_size,
+                          jentry->key_size};
+    };
+
+    auto entry_to_val = [](repl_journal_entry* jentry) {
+        return sisl::blob{uintptr_cast(jentry) + sizeof(repl_journal_entry) + jentry->user_header_size +
+                              jentry->key_size,
+                          jentry->value_size};
+    };
+
+    repl_key const rkey{.server_id = jentry->server_id, .term = lentry->get_term(), .dsn = jentry->dsn};
+
+    auto const [it, happened] = m_repl_key_req_map.try_emplace(rkey, repl_req_ptr_t(new repl_req_ctx()));
+    RD_DBG_ASSERT((it != m_repl_key_req_map.end()), "Unexpected error in map_repl_key_to_req");
+    auto rreq = it->second;
+    RD_DBG_ASSERT(happened, "rreq already exists for rkey={}", rkey.to_string());
+    MultiBlkId entry_blkid;
+    entry_blkid.deserialize(entry_to_val(jentry), true /* copy */);
+    rreq->set_local_blkid(entry_blkid);
+    rreq->init(rkey, jentry->code, false /* is_proposer */, entry_to_hdr(jentry), entry_to_key(jentry),
+               (entry_blkid.blk_count() * get_blk_size()));
```

Comment:
Nit set_local_blkid could be done after init?

Reply:
sounds good